### PR TITLE
fix: fix RegisterOnChange methods for journal and db_slice

### PR DESCRIPTION
**Pull request by BorysTheDev**
We have a bug with the journal->RegisterOnChange functions:

1) Start 2 tasks with traversing through the whole DB
2) The first task is already work
3) Second task do db_slice_->RegisterOnChange(db_callback)
4) After that second task do journal->RegisterOnChange(..) and yield to another fiber on the mutex JournalSlice::cb_mu_
5) during the 4th step we do LPUSH for example and db_callback save the bucket
6) we still don't have a callback for the journal so we lost data for LPUSH
7) we finish journal->RegisterOnChange(..)

This issue was reproducible in test_cluster_fuzzymigration

Also, I have fixed the issue with db_slice_->UnregisterOnChange() if it was called during callbacks processing (if the fiber yields during callback)

### CODE REVIEW

**Comment by romange** [comment]
Please provide a PR description. In general, I would like to ask to provide a better context for non -trivial bug fixes. See https://github.com/dragonflydb/dragonfly/pull/3103 for example how @adiholden  does it.

**Comment by romange** [comment]
@BorysTheDev  can you also explain the fix - i.e. why shared_ptr solves the issue?

**Comment by BorysTheDev** [comment]
@romange the problem was in mutex that could yield to another fiber but prevented removing a callback during a call. So to fix the issue I need to remove mutex, in this case, the fiber doesn't yield, but I need to prevent removing a callback so I created a copy of shared_ptr

File: src/server/db_slice.cc
Comment (dranikpg) [inline src/server/db_slice.cc:464]: maybe I don't understand something, but modifications while iterating are still unsafe. The vector might re-allocate and all your iterators become invalid (the ones in the for loop)

Comment (BorysTheDev) [reply src/server/db_slice.cc:464]: You are right. I also have found one more scenario that can crash. I will fix them tomorrow

**Comment by BorysTheDev** [comment]
to not forget, It looks we need mutex too prevent removing some data while callback is in progress. Will think about better fix tomorrow

File: src/server/journal/journal_slice.cc
Comment (BorysTheDev) [reply src/server/journal/journal_slice.cc:195]: forget to remove

File: src/server/journal/journal_slice.h
Comment (BorysTheDev) [reply src/server/journal/journal_slice.h:64]: it is used to get size of the callbacks vector during callbacks call

Comment (BorysTheDev) [reply src/server/journal/journal_slice.h:64]: I don't think that it really helps for multiple fibers, but if you want I can return it

Comment (romange) [reply src/server/journal/journal_slice.h]: I did not ask, i just asked :)

Comment (BorysTheDev) [reply src/server/journal/journal_slice.h]: I got it. And thought a little better and I decided that I was wrong)

**Comment by romange** [comment]
LGTM from my side, will wait for Adi to approve

**Comment by adiholden** [comment]
I am not sure about the correctness of this solution.
In this change we dont have a mutex in JournalSlice::RegisterOnChange

Lets take for example the following scenario:
1. callback "LPUSH key val" is executed
2. we run db_slice callbacks for bucket holding the key
3. we give new version to the bucket holding key - version x
4. we run journal_slice callback to write journal change "lpush key val"
5. we yield when writing the journal change 
6. another replica is starting calling db_slice register on change with version x+1
7. journal slice is registered on change
8. we return to fiber yield in step 5 continue iterating the journal changes list
9. we no write the journal change "lpush key val" to the stream before we serialized the bucket

File: src/server/db_slice.h
Comment (adiholden) [inline src/server/db_slice.h:549]: can you explain the comment?
during call of what?
currently we do not allow preemption in change_cb_, if they preempt we have a bug probably

File: src/server/journal/journal_slice.cc
Comment (adiholden) [inline src/server/journal/journal_slice.cc:194]: I cant be sure that this solution is not with bugs
If we have some place where preempt between callback execution and the call to AddLogRecord this solution still has the bug.

What we want is to make the register to dbsilce and journal atomic, so what we can do is lock before transaction callback is executed (Transaction::RunCallback) and release the lock after we write the journal change LogAutoJournalOnShard

Comment (BorysTheDev) [reply src/server/journal/journal_slice.cc:194]: You FiberAtomicGuard?

File: src/server/db_slice.h
Comment (BorysTheDev) [reply src/server/db_slice.h:549]: you are right we have FiberAtomicGuard there

Comment (BorysTheDev) [reply src/server/db_slice.h:549]: You are right. We don't have a bug

File: src/server/journal/journal_slice.cc
Comment (adiholden) [inline src/server/journal/journal_slice.cc:193]: do we still need this?

**Comment by adiholden** [comment]
@BorysTheDev The changes looks good. Not sure why he ci failed here.
Also please make sure to run regression tests on you branch before merging

**Comment by kostasrim** [comment]
> @BorysTheDev The changes looks good. Not sure why he ci failed here. Also please make sure to run regression tests on you branch before merging

yes and I will be merging my CI changes and we have a failure on cluster slot which I disabled manually -- keep that in mind as well

File: src/server/detail/save_stages_controller.cc
Comment (dranikpg) [inline src/server/detail/save_stages_controller.cc:300]: So what do we do with multi-hop operations and multi commands now? We can start serializing non-atomic values? There surely was some reason that SAVE was transctional

File: .vscode/launch.json
Comment (adiholden) [inline .vscode/launch.json:8]: revert

File: src/server/db_slice.h
Comment (adiholden) [inline src/server/db_slice.h:559]: Please add this comment, you can remove the comments above LockChangeCb/UnlockChangeCb
To ensure correct data replication, we must serialize the buckets that each running command will modify, followed by serializing the command to the journal. We use a mutex to prevent interleaving between bucket and journal registrations, and the command execution with its journaling. LockChangeCb is called before the callback, and UnlockChangeCb is called after journaling is completed. Register to bucket and journal changes is also does without preemption

**Review by adiholden — APPROVED** [review]

**Pull request merged**
